File: core/execution/simulator.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Tuple

from core.models.snapshots import MarketSnapshot
from strategies.base import StrategyAction


@dataclass
class SimulatedFill:
    strategy_id: str
    kalshi_ticker: str
    action_type: str
    side: str | None
    size: int | None
    price: float | None
    reason: str | None
    meta: Dict[str, Any]


@dataclass
class Position:
    size: int = 0
    avg_price: float = 0.0
    realized_pnl: float = 0.0


class ExecutionSimulator:
    def __init__(self, config: Dict[str, Any]) -> None:
        self.config = config

        self.contract_per_trade: int = int(self.config.get("contract_per_trade", 1))
        self.max_exposure_usd: float = float(self.config.get("max_exposure_usd", 0.0))
        self.slippage_buy: float = float(self.config.get("slippage_buy", 0.0))
        self.slippage_sell: float = float(self.config.get("slippage_sell", 0.0))

        self._positions: Dict[Tuple[str, str], Position] = {}
        self._strategy_exposure: Dict[str, float] = {}
        self._marks: Dict[Tuple[str, str], float] = {}

    def simulate_actions(
        self, strategy_id: str, actions: List[StrategyAction], snapshot: MarketSnapshot
    ) -> List[SimulatedFill]:
        fills: List[SimulatedFill] = []

        if not actions:
            return fills

        for action in actions:
            size = int(action.size or self.contract_per_trade)
            if size <= 0:
                continue

            if action.kalshi_ticker != snapshot.kalshi_ticker:
                continue

            fill = self._simulate_single_action(strategy_id, action, size, snapshot)
            if fill is not None:
                fills.append(fill)

        return fills

    def get_strategy_state(self, strategy_id: str) -> Dict[str, Any]:
        positions: Dict[str, Dict[str, Any]] = {}
        realized_pnl_total = 0.0
        unrealized_pnl_total = 0.0

        for (sid, ticker), pos in self._positions.items():
            if sid != strategy_id:
                continue

            key = (sid, ticker)
            mark = self._marks.get(key, pos.avg_price)

            unrealized_pnl = pos.size * (mark - pos.avg_price)

            positions[ticker] = {
                "size": pos.size,
                "avg_price": pos.avg_price,
                "realized_pnl": pos.realized_pnl,
                "mark": mark,
                "unrealized_pnl": unrealized_pnl,
            }

            realized_pnl_total += pos.realized_pnl
            unrealized_pnl_total += unrealized_pnl

        exposure = self._get_exposure(strategy_id)
        return {
            "strategy_id": strategy_id,
            "state": {
                "positions": positions,
                "exposure": exposure,
                "realized_pnl": realized_pnl_total,
                "unrealized_pnl": unrealized_pnl_total,
            },
        }

    def update_mark(self, strategy_id: str, snapshot: MarketSnapshot) -> None:
        ticker = snapshot.kalshi_ticker
        if not ticker:
            return

        key = (strategy_id, ticker)
        pos = self._positions.get(key)
        if not pos or pos.size == 0:
            return

        mid = snapshot.kalshi_yes_mid
        bid = snapshot.kalshi_yes_bid
        ask = snapshot.kalshi_yes_ask

        mark: Optional[float] = None
        if mid is not None:
            mark = mid
        elif bid is not None and ask is not None:
            mark = 0.5 * (bid + ask)
        elif bid is not None:
            mark = bid
        elif ask is not None:
            mark = ask

        if mark is None:
            return

        mark = max(0.0, min(1.0, mark))
        self._marks[key] = mark

    def _get_position(self, strategy_id: str, ticker: str) -> Position:
        key = (strategy_id, ticker)
        pos = self._positions.get(key)
        if pos is None:
            pos = Position()
            self._positions[key] = pos
        return pos

    def _get_exposure(self, strategy_id: str) -> float:
        return float(self._strategy_exposure.get(strategy_id, 0.0))

    def _set_exposure(self, strategy_id: str, value: float) -> None:
        self._strategy_exposure[strategy_id] = float(value)

    def _simulate_single_action(
        self,
        strategy_id: str,
        action: StrategyAction,
        size: int,
        snapshot: MarketSnapshot,
    ) -> Optional[SimulatedFill]:
        ticker = action.kalshi_ticker
        pos = self._get_position(strategy_id, ticker)
        pre_size = pos.size
        pre_exposure = self._get_exposure(strategy_id)

        side = action.side
        if side not in ("BUY", "SELL"):
            return None

        price = self._compute_fill_price(side, snapshot)
        if price is None:
            return None

        if side == "SELL" and size > pos.size:
            size = pos.size
            if size <= 0:
                return None

        if side == "BUY" and self.max_exposure_usd > 0.0:
            notional = size * price
            projected_exposure = pre_exposure + notional
            if projected_exposure > self.max_exposure_usd:
                return SimulatedFill(
                    strategy_id=strategy_id,
                    kalshi_ticker=ticker,
                    action_type=action.type,
                    side=side,
                    size=0,
                    price=None,
                    reason=action.reason,
                    meta={
                        "rejected": True,
                        "rejected_reason": "max_exposure_exceeded",
                        "pre_exposure": pre_exposure,
                        "max_exposure_usd": self.max_exposure_usd,
                    },
                )

        if side == "BUY":
            pos = self._apply_buy(pos, size, price)
            post_exposure = pre_exposure + size * price
        else:
            realized_delta, new_size = self._apply_sell(pos, size, price)
            pos.size = new_size
            pos.realized_pnl += realized_delta
            post_exposure = max(pre_exposure - size * pos.avg_price, 0.0)

        self._set_exposure(strategy_id, post_exposure)

        return SimulatedFill(
            strategy_id=strategy_id,
            kalshi_ticker=ticker,
            action_type=action.type,
            side=side,
            size=size,
            price=price,
            reason=action.reason,
            meta={
                "pre_size": pre_size,
                "post_size": pos.size,
                "pre_exposure": pre_exposure,
                "post_exposure": post_exposure,
                "realized_pnl": pos.realized_pnl,
            },
        )

    def _compute_fill_price(
        self,
        side: str,
        snapshot: MarketSnapshot,
    ) -> Optional[float]:
        mid = snapshot.kalshi_yes_mid
        bid = snapshot.kalshi_yes_bid
        ask = snapshot.kalshi_yes_ask

        if side == "BUY":
            base = ask if ask is not None else mid
            if base is None:
                return None
            price = base + self.slippage_buy
        else:
            base = bid if bid is not None else mid
            if base is None:
                return None
            price = base - self.slippage_sell

        return max(0.0, min(1.0, price))

    def _apply_buy(self, pos: Position, size: int, price: float) -> Position:
        total_size = pos.size + size
        if total_size <= 0:
            pos.size = 0
            pos.avg_price = 0.0
            return pos

        new_notional = pos.size * pos.avg_price + size * price
        pos.size = total_size
        pos.avg_price = new_notional / total_size
        return pos

    def _apply_sell(self, pos: Position, size: int, price: float) -> tuple[float, int]:
        if size <= 0 or pos.size <= 0:
            return 0.0, pos.size

        close_size = min(size, pos.size)
        realized = close_size * (price - pos.avg_price)
        new_size = pos.size - close_size

        if new_size == 0:
            pos.avg_price = 0.0

        return realized, new_size

File: src-tauri/src/engine/simulator.rs
#![allow(dead_code)]

use super::models::{MarketSnapshot, Position, SimulatedFill, StrategyAction};
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct ExecutionSimulator {
    pub contract_per_trade: i64,
    pub max_exposure_usd: f64,
    pub slippage_buy: f64,
    pub slippage_sell: f64,
    positions: HashMap<(String, String), Position>,
    marks: HashMap<(String, String), f64>,
    exposures: HashMap<String, f64>,
}

impl ExecutionSimulator {
    pub fn new(contract_per_trade: i64, max_exposure_usd: f64, slippage_buy: f64, slippage_sell: f64) -> Self {
        Self {
            contract_per_trade,
            max_exposure_usd,
            slippage_buy,
            slippage_sell,
            positions: HashMap::new(),
            marks: HashMap::new(),
            exposures: HashMap::new(),
        }
    }

    pub fn update_mark(&mut self, strategy_id: &str, snapshot: &MarketSnapshot) {
        let key = (strategy_id.to_string(), snapshot.kalshi_ticker.clone());
        let mark = snapshot
            .kalshi_yes_mid
            .or_else(|| match (snapshot.kalshi_yes_bid, snapshot.kalshi_yes_ask) {
                (Some(b), Some(a)) => Some(0.5 * (b + a)),
                (Some(b), None) => Some(b),
                (None, Some(a)) => Some(a),
                _ => None,
            })
            .unwrap_or(0.50)
            .clamp(0.01, 0.99);

        self.marks.insert(key.clone(), mark);

        if let Some(pos) = self.positions.get_mut(&key) {
            if pos.size != 0 {
                pos.mark = mark;
                pos.unrealized_pnl = pos.size as f64 * (mark - pos.avg_price);
            }
        }
    }

    pub fn simulate_actions(
        &mut self,
        strategy_id: &str,
        actions: &[StrategyAction],
        snapshot: &MarketSnapshot,
    ) -> Vec<SimulatedFill> {
        let mut fills = Vec::new();

        for action in actions {
            if action.kalshi_ticker != snapshot.kalshi_ticker {
                continue;
            }

            let side = match action.side.as_deref() {
                Some("BUY") => "BUY",
                Some("SELL") => "SELL",
                _ => continue,
            };

            let req_size = action.size.unwrap_or(self.contract_per_trade);
            if req_size <= 0 {
                continue;
            }

            let key = (strategy_id.to_string(), snapshot.kalshi_ticker.clone());
            let pos = self.positions.entry(key.clone()).or_insert_with(|| Position {
                instrument: snapshot.kalshi_ticker.clone(),
                strategy_id: strategy_id.to_string(),
                size: 0,
                avg_price: 0.0,
                realized_pnl: 0.0,
                mark: snapshot.kalshi_yes_mid.unwrap_or(0.50),
                unrealized_pnl: 0.0,
            });

            let price = match side {
                "BUY" => {
                    let base = snapshot.kalshi_yes_ask.or(snapshot.kalshi_yes_mid);
                    match base {
                        Some(p) => (p + self.slippage_buy).clamp(0.01, 0.99),
                        None => continue,
                    }
                }
                "SELL" => {
                    let base = snapshot.kalshi_yes_bid.or(snapshot.kalshi_yes_mid);
                    match base {
                        Some(p) => (p - self.slippage_sell).clamp(0.01, 0.99),
                        None => continue,
                    }
                }
                _ => continue,
            };

            let mut fill_size = req_size;
            if side == "SELL" && fill_size > pos.size {
                fill_size = pos.size;
                if fill_size <= 0 {
                    continue;
                }
            }

            let exposure = *self.exposures.get(strategy_id).unwrap_or(&0.0);
            if side == "BUY" && self.max_exposure_usd > 0.0 {
                let notional = fill_size as f64 * price;
                if exposure + notional > self.max_exposure_usd {
                    continue;
                }
            }

            if side == "BUY" {
                let total_size = pos.size + fill_size;
                let new_notional = pos.size as f64 * pos.avg_price + fill_size as f64 * price;
                pos.size = total_size;
                pos.avg_price = if total_size > 0 { new_notional / total_size as f64 } else { 0.0 };
                self.exposures.insert(strategy_id.to_string(), exposure + fill_size as f64 * price);
            } else {
                let close_size = fill_size.min(pos.size);
                let realized = close_size as f64 * (price - pos.avg_price);
                pos.size -= close_size;
                pos.realized_pnl += realized;
                if pos.size == 0 {
                    pos.avg_price = 0.0;
                }
                let new_exposure = (exposure - fill_size as f64 * pos.avg_price).max(0.0);
                self.exposures.insert(strategy_id.to_string(), new_exposure);
            }

            let mark = snapshot.kalshi_yes_mid.unwrap_or(price);
            pos.mark = mark;
            pos.unrealized_pnl = pos.size as f64 * (mark - pos.avg_price);
            self.marks.insert(key, mark);

            fills.push(SimulatedFill {
                strategy_id: strategy_id.to_string(),
                kalshi_ticker: snapshot.kalshi_ticker.clone(),
                action_type: action.action_type.clone(),
                side: side.to_string(),
                size: fill_size,
                price,
                reason: action.reason.clone(),
            });
        }

        fills
    }

    pub fn get_position(&self, strategy_id: &str, ticker: &str) -> Option<Position> {
        let key = (strategy_id.to_string(), ticker.to_string());
        self.positions.get(&key).cloned()
    }

    pub fn get_all_positions(&self, strategy_id: &str) -> Vec<Position> {
        self.positions
            .iter()
            .filter(|((sid, _), _)| sid == strategy_id)
            .map(|(_, pos)| {
                let key = (pos.strategy_id.clone(), pos.instrument.clone());
                let mark = *self.marks.get(&key).unwrap_or(&pos.avg_price);
                let mut p = pos.clone();
                p.mark = mark;
                p.unrealized_pnl = p.size as f64 * (mark - p.avg_price);
                p
            })
            .collect()
    }

    pub fn get_total_pnl(&self, strategy_id: Option<&str>) -> f64 {
        let mut total = 0.0;
        for ((sid, _), pos) in &self.positions {
            if let Some(target) = strategy_id {
                if sid != target {
                    continue;
                }
            }
            let key = (sid.clone(), pos.instrument.clone());
            let mark = *self.marks.get(&key).unwrap_or(&pos.avg_price);
            let unrealized = pos.size as f64 * (mark - pos.avg_price);
            total += pos.realized_pnl + unrealized;
        }
        total
    }
}
